File: Furthest Building You Can Reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        int n = heights.size();
        vector<int> previous;
        int difference;
        int counter = 0;
        int i;
        
        if(n-1 <= ladders)
            return n-1;
        
        for(i = 0 ; i<n-1; i++)
        {
            if(heights[i] >= heights[i+1])
            {
                counter ++;
                continue;
            }
            
            difference = heights[i+1] - heights[i];
            previous.push_back(difference);
            push_heap(previous.begin(),previous.end(),greater());
            
            if(previous[0] > bricks && previous.size() > ladders)
                break;
            
            if(previous.size() > ladders)
            {
                if(previous[0] <= bricks)
                {
                    counter ++;
                    bricks -= previous[0];
                    pop_heap(previous.begin(),previous.end(),greater());
                    previous.pop_back();
                }
            }
            
        }
        
        if(n-1 < counter + ladders)
            return n-1;
        else
             return counter + ladders;
        
    }
};

File: furthest-building-you-can-reach.py
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        bricks_used = []
        
        for i in range(1,len(heights)):
            diff = heights[i] - heights[i-1]
            if  diff > 0:
                if bricks - diff >= 0:
                    bricks -= diff
                    heappush(bricks_used,-diff)
                else:
                    if ladders == 0:
                        return i - 1
                    if len(bricks_used) and -bricks_used[0] >= diff:
                        bricks += -heappop(bricks_used)
                        bricks -= diff
                        heappush(bricks_used, -diff)
                    ladders -= 1

        return len(heights) - 1